tracing-attributes/src/lib.rs

```diff
@@ -835,125 +893,155 @@ mod kw {
+            // If the function is async, this is a candidate
             if fun.sig.asyncness.is_some() {
-                inside_funs.push(fun);
+                return Some((stmt, fun));
             }
-        } else if let Stmt::Expr(e) = &stmt {
-            last_expr = Some(e);
         }
-    }
+        None
+    });
 
-    // let's play with (too much) pattern matching
-    // is the last expression a function call?
-    if let Some(Expr::Call(ExprCall {
-        func: outside_func,
-        args: outside_args,
-        ..
-    })) = last_expr
-    {
-        if let Expr::Path(path) = outside_func.as_ref() {
-            // is it a call to `Box::pin()`?
-            if "Box::pin" == path_to_string(&path.path) {
-                // does it takes at least an argument? (if it doesn't,
-                // it's not gonna compile anyway, but that's no reason
-                // to (try to) perform an out of bounds access)
-                if outside_args.is_empty() {
-                    return None;
-                }
-                // is the argument to Box::pin a function call itself?
-                if let Expr::Call(ExprCall { func, .. }) = &outside_args[0] {
-                    if let Expr::Path(inside_path) = func.as_ref() {
-                        // "stringify" the path of the function called
-                        let func_name = path_to_string(&inside_path.path);
-                        // is this function directly defined insided the current block?
-                        for fun in inside_funs {
-                            if fun.sig.ident == func_name {
-                                // we must hook this function now
-                                return Some(fun);
-                            }
-                        }
-                    }
-                }
-            }
+    // last expression of the block (it determines the return value
+    // of the block, so that if we are working on a function whose
+    // `trait` or `impl` declaration is annotated by async_trait,
+    // this is quite likely the point where the future is pinned)
+    let (last_expr_stmt, last_expr) = block.stmts.iter().rev().find_map(|stmt| {
+        if let Stmt::Expr(expr) = stmt {
+            Some((stmt, expr))
+        } else {
+            None
         }
+    })?;
+
+    // is the last expression a function call?
+    let (outside_func, outside_args) = match last_expr {
+        Expr::Call(ExprCall { func, args, .. }) => (func, args),
+        _ => return None,
+    };
+
+    // is it a call to `Box::pin()`?
+    const BOX_PIN_NAMES: [&str; 6] = [
+        "Box::pin",
+        "std::boxed::Box::pin",
+        "::std::boxed::Box::pin",
+        "alloc::boxed::Box::pin",
+        "::alloc::boxed::Box::pin",
+        "boxed::Box::pin",
+    ];
+    let path = match outside_func.as_ref() {
+        Expr::Path(path) => &path.path,
+        _ => return None,
+    };
+    if !BOX_PIN_NAMES.contains(&path_to_string(path).as_str()) {
```

Comment:
it still seems like it might be better just to write
```suggestion
    // is it a call to `Box::pin()`?
    let path = match outside_func.as_ref() {
        Expr::Path(path) => &path.path,
        _ => return None,
    };
    if &path_to_string(path).ends_with("Box::pin") {
```
since this will also deal with cases where `Box::pin` is reexported under some other path (unlikely, but some users in weird use-cases might do this). also, it's probably faster --- it's one string comparison, rather than having to iterate the whole array, and it's simpler.

what do you think?

Reply:
Yes, it's simpler, and there is probably no conflict with user-defined functions `pin` in a namespace `Box` that do not have the same semantics as the "true" `Box::pin` (I sure hopes nobody does this).